source/git/gitsource_unix.go

```diff
@@ -4,26 +4,83 @@ package git
 
 import (
 	"context"
+	"os"
 	"os/exec"
-	"syscall"
+	"os/signal"
 	"time"
+
+	"github.com/docker/docker/pkg/reexec"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	gitCmd = "umask-git"
 )
 
+func init() {
+	reexec.Register(gitCmd, gitMain)
+}
+
+func gitMain() {
+	// Need standard user umask for git process.
+	unix.Umask(0022)
+
+	// Reexec git command
+	cmd := exec.Command(os.Args[1], os.Args[2:]...)
```

Comment:
btw, `syscall.Exec` also exists. Not sure if there could be problems though. Eg. eventual windows support maybe.

Reply:
Found the issue actually for the infinite spawned process with reexec. Was because `/proc/self/exe` through `go test` doesn't init reexec.